cm/master.py

```diff
@@ -1463,6 +1386,111 @@ def _add_data_fs(fs_name=None):
+            for template in [template for template in self.app.config.cluster_templates if 'name' in template]:
+                if template['name'] == cluster_type:
+                    cluster_template = template
+            if not cluster_template or (not 'filesystem_templates' in cluster_template and cluster_type != "Test"):
+                log.warn("No filesystem templates defined for cluster type: {0}".format(cluster_type))
+            else:
+                self.process_filesystem_templates(cluster_type, pss, storage_type, cluster_template['filesystem_templates'])
+        elif self.app.config.filesystem_templates: # Legacy behavior - process filesystem_templates directly or snaps.yaml
+            self.process_filesystem_templates(cluster_type, pss, storage_type, self.app.config.filesystem_templates)
+        else:
+            if cluster_type != "Test":
+                log.error("No filesystem templates defined for cluster type: {0}".format(cluster_type))
+
+    # TODO: Remove special case handling for Galaxy filesystem - requires change in UI and initialisation logic
+    def process_filesystem_templates(self, cluster_type, pss, storage_type, filesystem_templates):
+        # Turn those data sources into file systems
+        if filesystem_templates:
+            attached_volumes = self.get_attached_volumes()
+            for fs_template in [s for s in filesystem_templates if 'name' in s]:
+                log.debug("Processing filesystem template: {0}".format(fs_template['name']))
+                fs = Filesystem(self.app, fs_template['name'],
+                                svc_roles=ServiceRole.from_string_array(fs_template.get('roles', None)))
+                # Check if an already attached volume maps to the current filesystem
+                att_vol = self.get_vol_if_fs(attached_volumes, fs_template['name'])
+                if att_vol:
+                    log.debug("{0} file system has volume(s) already attached".format(
+                        fs_template['name']))
+                    fs.add_volume(vol_id=att_vol.id,
+                                  size=att_vol.size, from_snapshot_id=att_vol.snapshot_id)
+                elif 'snap_id' in fs_template:
+                    log.debug("There are no volumes already attached for file system {0}"
+                              .format(fs_template['name']))
+                    size = max(fs_template.get('size', 0), fs_template.get('min_size', 0)) # always default to 0 for snaps
+                    if ServiceRole.GALAXY_DATA in ServiceRole.from_string_array(fs_template.get('roles', None)):
+                        size = pss
+                    fs.add_volume(size=size, from_snapshot_id=fs_template['snap_id'])
+                elif 'type' in fs_template:
+                    # TODO: This obviates the need for type archive. Instead, archive should be a volume "data_source"
+                    if 'archive' == fs_template['type'] and 'archive_url' in fs_template:
```

Comment:
Yes, the data needs to be in yaml format, and flavours are available in cloudlaunch admin under images. Each image can have multiple flavours defined. An example is as follows:

```
bucket_default: cloudman-gvl-400`
initial_cluster_type: Galaxy
galaxy_data_option: transient
post_start_script_url: 'file:///opt/gvl/scripts/triggers;file:///mnt/galaxy/gvl/poststart.d'
cluster_templates:
  - name: Galaxy
    filesystem_templates:
    - name: galaxy
      type: transient
      roles: galaxyTools,galaxyData
      data_source: archive
      archive_url: https://swift.rc.nectar.org.au:8888/v1/AUTH_377/cloudman-os/gvl-galaxyfs-4.0.0.tar.gz
      archive_md5: cccf12df2cf4c9f3d9af6526fee1bbfc
    - name: galaxyIndices
      type: volume
      roles: galaxyIndices
      data_source: archive
      archive_url: https://swift.rc.nectar.org.au:8888/v1/AUTH_377/cloudman-dev/galaxyIndices-3.05.tar.gz
      archive_md5: 32f7b4c896937236b7ac862090adb3df
  - name: Data
    filesystem_templates:
    - name: galaxy
      type: volume
```